libs/utils/trace.py

```diff
@@ -886,6 +886,69 @@ def _compute_steps_duration(self, df):
 
         return df
 
+    def _windowify_steps_df(self, df, (start, end), len_column='len'):
+        """
+        Slice a dataframe to [start:end], and work on the time data so that it
+        makes sense within the interval.
+
+        Examples to make it clearer:
+
+        df is:
+        Time len state
+        15    1   1
+        16    1   0
+        17    1   1
+        18    1   0
+        -------------
+
+        _windowify_steps_df(df, (16.5, 17.5)) =>
+
+        Time len state
+        16.5  .5   0
+        17    .5   1
+
+        _windowify_steps_df(df, (16.2, 16.8)) =>
+
+        Time len state
+        16.2  .6   0
+
+        :returns: a new df that fits the above description
+        """
+        if df.empty:
+            return df
+
+        res = df[start:end].copy()
+
+        # There's a 99% probability that events won't fit snuggly within within
+        # the interval. Find the one preceding the interval, and deduce the rest
+        # mathemagically.
+        left = df[:start].tail(1)
+        right = res.tail(1)
+
+        # "void" to fill between 'start' and the first event in the interval
+        if not left.empty:
+            left_duration = max(0, left.index[0] + left[len_column].values[0] - start)
```

Comment:
Well for normally shaped dfs I would say it could work.The example dataframe has continuous events... But what if it doesn't ? That's totally the case for my use of it: I take the latency_df and filter events that I don't care about (I only keep the active state). So I end up with a dataframe that has a 'len' column but those are not continuous.

This was all very clear to me when I wrote it but now it's really confusing tbh.

Reply:
Ok, I was missing this use-case... then probably also the solution below does not fits that use-case.

However, I don't see why you need to filter (i.e. keep only active states) and then trim after...